faunadb-common/src/main/java/com/faunadb/common/Connection.java

```diff
@@ -112,57 +127,80 @@ public Builder withMetrics(MetricRegistry registry) {
-        r = new MetricRegistry();
+        registry = new MetricRegistry();
       else
-        r = metricRegistry;
+        registry = metricRegistry;
 
-      AsyncHttpClient c;
+      AsyncHttpClient httpClient;
       if (client == null) {
-        AsyncHttpClientConfig config = new DefaultAsyncHttpClientConfig.Builder()
-          .setConnectTimeout(DEFAULT_CONNECTION_TIMEOUT_MS)
-          .setRequestTimeout(DEFAULT_REQUEST_TIMEOUT_MS)
-          .setPooledConnectionIdleTimeout(DEFAULT_IDLE_TIMEOUT_MS)
-          .setMaxRequestRetry(0)
-          .build();
-        c = new DefaultAsyncHttpClient(config);
-      } else
-        c = client;
+        httpClient = new DefaultAsyncHttpClient(
+          new DefaultAsyncHttpClientConfig.Builder()
+            .setConnectTimeout(DEFAULT_CONNECTION_TIMEOUT_MS)
+            .setRequestTimeout(DEFAULT_REQUEST_TIMEOUT_MS)
+            .setPooledConnectionIdleTimeout(DEFAULT_IDLE_TIMEOUT_MS)
+            .setMaxRequestRetry(DEFAULT_MAX_RETRIES)
+            .build()
+        );
+      } else {
+        httpClient = client;
+      }
 
       URL root;
       if (faunaRoot == null)
-        root = new URL("https://rest.faunadb.com");
+        root = FAUNA_ROOT;
       else
         root = faunaRoot;
 
-      return new Connection(root, authToken, c, r);
+      return new Connection(root, authToken, httpClient, registry, new AtomicInteger(1));
     }
   }
 
-  private static String XFaunaDBHost = "X-FaunaDB-Host";
-  private static String XFaunaDBBuild = "X-FaunaDB-Build";
+  private static final String ASCII = "ASCII";
+  private static final String X_FAUNADB_HOST = "X-FaunaDB-Host";
+  private static final String X_FAUNADB_BUILD = "X-FaunaDB-Build";
 
   private final URL faunaRoot;
-  private final String authToken;
   private final String authHeader;
   private final AsyncHttpClient client;
   private final MetricRegistry registry;
+  private final AtomicInteger refCount;
 
   private final Logger log = LoggerFactory.getLogger(getClass());
   private final ObjectMapper json = new ObjectMapper();
 
-  Connection(URL faunaRoot, String authToken, AsyncHttpClient client, MetricRegistry registry) throws UnsupportedEncodingException {
+  private Connection(URL faunaRoot, String authToken, AsyncHttpClient client, MetricRegistry registry, AtomicInteger refCount) {
     this.faunaRoot = faunaRoot;
-    this.authToken = authToken;
     this.authHeader = generateAuthHeader(authToken);
     this.client = client;
     this.registry = registry;
+    this.refCount = refCount;
+  }
+
+  /**
+   * Creates a new short live connection sharing its underneath {@link AsyncHttpClient}. Queries submited to a
+   * session connection will be authenticated with the token provided.
+   *
+   * @param authToken token that will be used to authenticate requests
+   * @return a new {@link Connection}
+   */
+  public Connection newSessionConnection(String authToken) {
+    if (refCount.incrementAndGet() > 1)
```

Comment:
This is how the driver API looks like if we pass a function rather returning a new client instance:

```diff
-    try (FaunaClient sessionClient = client.newSessionClient(secret)) {
-      Value loggedOut = sessionClient.query(Logout(Value(true))).get();
-      assertThat(loggedOut.to(BOOLEAN).get(), is(true));
-    }
+    Value loggedOut = client.newSessionWith(secret, new Function<FaunaClient, ListenableFuture<Value>>() {
+      @Override
+      public ListenableFuture<Value> apply(FaunaClient client) {
+        return client.query(Logout(Value(true)));
+      }
+    }).get();
+
+    assertThat(loggedOut.to(BOOLEAN).get(), is(true));
```

It's kind of a silly test but, it illustrates the behaviour.

Although is more verbose, it has the advantage of being a single method call. We still need to control `refCount` internally because `close` is a public method and anyone could call it but, we can call `close` after the session function is applied, which give us more control and the user doesn't need to worry about calling it.

There is always the change that the user can leak the inner `FaunaClient` out of the function scope but, the function closure makes it more difficult (and clearly wrong) than previous version.

Any thoughts?